cli/src/main/java/com/box/l10n/mojito/cli/command/checks/SpellCliChecker.java

```diff
@@ -0,0 +1,228 @@
+            stringWithoutPlaceholders = stringWithoutPlaceholders.replaceAll(tmp, "");
+            index = stringWithoutPlaceholders.indexOf("{");
+        }
+        return stringWithoutPlaceholders;
+    }
+
+    private int getEndOfPlaceholderIndex(String str, int startIndex){
+        ArrayDeque<Character> stack = new ArrayDeque<>();
+        int indexCount = startIndex;
+        for (Character c : str.substring(startIndex).toCharArray()) {
+            if (c.equals('{')) {
+                stack.push(c);
+                indexCount++;
+                continue;
+            } else if (c.equals('}')) {
+                if(stack.isEmpty()){
+                    throw new CommandException("Invalid number of opening brackets in string.");
+                }
+                stack.pop();
+                if(stack.isEmpty()) {
+                    return indexCount;
+                }
+            }
+            indexCount++;
+        }
+        if(!stack.isEmpty()) {
+            throw new CommandException("Invalid number of closing brackets in string.");
+        }
+        return -1;
+    }
+
+    private void loadAdditionalWordsToDictionary(String additionalWordsFilePath) {
+
+        if (additionalWordsFilePath != null && !additionalWordsFilePath.isEmpty()) {
+            try {
+                Files.readAllLines(Paths.get(additionalWordsFilePath)).stream().forEach(word -> {
+                    String w = word.trim();
+                    logger.debug("Adding {} to dictionary", w);
+                    hunspell.add(w);
+                });
+            } catch (IOException e) {
+                logger.error("Error adding additional words to dictionary: {}", e);
+                throw new CommandException("Error adding additional words to dictionary: " + e.getMessage());
+            }
+        }
+    }
+
+    private Map<String, List<String>> spellCheckSourceString(List<String> words) {
+        Map<String, List<String>> failureMap = new HashMap<>();
+        Set<String> checked = new HashSet<>();
+        for (String word : words) {
+            if (!checked.contains(word) && !hunspell.spell(word)) {
+                List<String> suggestions = hunspell.suggest(word);
+                logger.debug("{} is spelt incorrectly. Suggested correct spellings are {}", word, suggestions);
+                failureMap.put(word, suggestions);
+                checked.add(word);
+            }
+        }
+        return failureMap;
+    }
+
+    private List<String> getSourceStringsFromDiff() {
+        List<String> sourceStrings = new ArrayList<>();
+        getAddedTextUnits().stream().forEach(assetExtractorTextUnit -> sourceStrings.add(assetExtractorTextUnit.getSource()));
+        return sourceStrings;
+    }
+
+    private String buildNotificationText(Map<String, Map<String, List<String>>> failureMap) {
```

Comment:
Should the notification builder be in its own class?

Reply:
Initially that was my intention, every checker would call a separate class to build up the failure notification text. However the way the different checkers display failures isn't uniform, for example the SpellChecker needs to have a way to display multiple errors in a single source string as its possible for there to be multiple misspellings in a single string and then it also needs to provide suggestions for those misspellings.

Alternatively the  Context & Comment checker has a one to one mapping with strings and failures so it doesn't follow the same text format as the spell checker which is why I decided to let each checker manage its own failure text .